[ISSUE] databricks_mount issue when cluster used to create mount no longer exists

### Configuration
```hcl
resource "databricks_mount" "this" {
  cluster_id = var.cluster_id
  name = "experiments"
  s3 {
    instance_profile = databricks_instance_profile.ds.id
    bucket_name      = aws_s3_bucket.this.bucket
  }
}
```

### Steps to Reproduce
What should have happened?

The mount was created with a cluster that does no longer exist.
So on a later run, we create a new cluster and change var.cluster_id to the new cluster id (**0518-144629-3y2xc1vr**)

### Expected Behavior

The databricks_mount resource should use the new cluster id to apply the changes

### Actual Behavior

We have the following error : 
```
Error: Cluster 0309-172025-zu2hs5vz does not exist
```
This id corresponds to the cluster does not exist anymore

After some check the remote state captures the cluster id used during the first apply and it seems to be the source of the error, as the resource checks if the cluster exists after refreshing the remote state

```
$ terraform state show xxxx

resource "databricks_mount" "mounts" {
    cluster_id      = "0309-172025-zu2hs5vz"
    encryption_type = "sse-kms"
    ...
}
```

To sort this out, I had to manually update the remote state to change the cluster id.


### Terraform and provider versions

Please paste the output of `terraform version`. If version of `databricks` provider is not the latest (https://github.com/databrickslabs/terraform-provider-databricks/releases), please make sure to use the latest one.

Terraform v1.1.9
Provider 0.5.7

### Debug Output

here is the interesting part

```
databricks_mount.mounts["/mnt/rd/datalake/raw"]: Refreshing state... [id=rd/datalake/raw]
databricks_mount.mounts["/mnt/rd/artefacts"]: Refreshing state... [id=rd/artefacts]
databricks_mount.mounts["/mnt/rd/logs"]: Refreshing state... [id=rd/logs]
2022-05-18T17:29:55.853+0200 [INFO]  provider.terraform-provider-databricks_v0.5.7: Using Bearer authentication from ~/.databrickscfg: timestamp=2022-05-18T17:29:55.853+0200
2022-05-18T17:29:55.853+0200 [INFO]  provider.terraform-provider-databricks_v0.5.7: Configured databricks-cli auth: host=***REDACTED***, token=***REDACTED***, profile=***REDACTED***: timestamp=2022-05-18T17:29:55.853+0200
2022-05-18T17:29:55.853+0200 [DEBUG] provider.terraform-provider-databricks_v0.5.7: GET /api/2.0/clusters/get: timestamp=2022-05-18T17:29:55.853+0200
2022-05-18T17:29:55.920+0200 [DEBUG] provider.terraform-provider-databricks_v0.5.7: GET /api/2.0/clusters/get: timestamp=2022-05-18T17:29:55.920+0200
2022-05-18T17:29:55.988+0200 [DEBUG] provider.terraform-provider-databricks_v0.5.7: GET /api/2.0/clusters/get: timestamp=2022-05-18T17:29:55.987+0200
2022-05-18T17:29:56.246+0200 [DEBUG] provider.terraform-provider-databricks_v0.5.7: 400 Bad Request {
2022-05-18T17:29:56.246+0200 [WARN]  provider.terraform-provider-databricks_v0.5.7: /api/2.0/clusters/get:400 - Cluster 0309-172025-zu2hs5vz does not exist https://docs.databricks.com/dev-tools/api/latest/clusters.html#get: timestamp=2022-05-18T17:29:56.246+0200
2022-05-18T17:29:56.247+0200 [ERROR] vertex "databricks_mount.mounts[\"/mnt/rd/artefacts\"]" error: Cluster 0309-172025-zu2hs5vz does not exist
2022-05-18T17:29:56.381+0200 [DEBUG] provider.terraform-provider-databricks_v0.5.7: 400 Bad Request {
2022-05-18T17:29:56.381+0200 [WARN]  provider.terraform-provider-databricks_v0.5.7: /api/2.0/clusters/get:400 - Cluster 0309-172025-zu2hs5vz does not exist https://docs.databricks.com/dev-tools/api/latest/clusters.html#get: timestamp=2022-05-18T17:29:56.381+0200
2022-05-18T17:29:56.382+0200 [ERROR] vertex "databricks_mount.mounts[\"/mnt/rd/logs\"]" error: Cluster 0309-172025-zu2hs5vz does not exist
2022-05-18T17:29:56.388+0200 [DEBUG] provider.terraform-provider-databricks_v0.5.7: 400 Bad Request {
2022-05-18T17:29:56.388+0200 [WARN]  provider.terraform-provider-databricks_v0.5.7: /api/2.0/clusters/get:400 - Cluster 0309-172025-zu2hs5vz does not exist https://docs.databricks.com/dev-tools/api/latest/clusters.html#get: timestamp=2022-05-18T17:29:56.388+0200
2022-05-18T17:29:56.388+0200 [ERROR] vertex "databricks_mount.mounts[\"/mnt/rd/datalake/raw\"]" error: Cluster 0309-172025-zu2hs5vz does not exist
2022-05-18T17:29:56.388+0200 [ERROR] vertex "databricks_mount.mounts" error: Cluster 0309-172025-zu2hs5vz does not exist
2022-05-18T17:29:56.388+0200 [ERROR] vertex "databricks_mount.mounts" error: Cluster 0309-172025-zu2hs5vz does not exist
2022-05-18T17:29:56.388+0200 [ERROR] vertex "databricks_mount.mounts" error: Cluster 0309-172025-zu2hs5vz does not exist
2022-05-18T17:29:56.388+0200 [ERROR] vertex "databricks_mount.mounts (expand)" error: Cluster 0309-172025-zu2hs5vz does not exist
```

--- a/storage/aws_s3_mount.go
+++ b/storage/aws_s3_mount.go
@@ -116,11 +116,7 @@ func preprocessS3Mount(ctx context.Context, d *schema.ResourceData, m interface{
 			return fmt.Errorf("cluster %s must have EC2 instance profile attached", clusterID)
 		}
 	} else if instanceProfile != "" {
-		cluster, err := GetOrCreateMountingClusterWithInstanceProfile(clustersAPI, instanceProfile)
-		if err != nil {
-			return err
-		}
-		return d.Set("cluster_id", cluster.ClusterID)
+		return mountS3ViaProfileAndSetClusterID(clustersAPI, instanceProfile, d)
 	}
 	return nil
 }

--- a/storage/aws_s3_mount_test.go
+++ b/storage/aws_s3_mount_test.go
@@ -5,8 +5,8 @@ import (
 	"testing"
 
 	"github.com/databrickslabs/terraform-provider-databricks/clusters"
-	"github.com/databrickslabs/terraform-provider-databricks/common"
 	"github.com/databrickslabs/terraform-provider-databricks/commands"
+	"github.com/databrickslabs/terraform-provider-databricks/common"
 
 	"github.com/databrickslabs/terraform-provider-databricks/qa"
 	"github.com/stretchr/testify/assert"
@@ -82,7 +82,7 @@ func TestResourceAwsS3MountCreate_invalid_arn(t *testing.T) {
 		},
 		Create: true,
 	}.Apply(t)
-	require.EqualError(t, err, "invalid arn: this_mount")
+	require.EqualError(t, err, "mount via profile: invalid arn: this_mount")
 }
 
 func TestResourceAwsS3MountRead(t *testing.T) {

--- a/storage/resource_mount_test.go
+++ b/storage/resource_mount_test.go
@@ -288,7 +288,7 @@ func TestResourceAwsS3MountGenericCreate_invalid_arn(t *testing.T) {
 		},
 		Create: true,
 	}.Apply(t)
-	require.EqualError(t, err, "invalid arn: this_mount")
+	require.EqualError(t, err, "mount via profile: invalid arn: this_mount")
 }
 
 func TestResourceAwsS3MountGenericRead(t *testing.T) {

--- a/storage/s3.go
+++ b/storage/s3.go
@@ -62,6 +62,12 @@ func preprocessS3MountGeneric(ctx context.Context, s map[string]*schema.Schema, 
 	clustersAPI := clusters.NewClustersAPI(ctx, m)
 	if clusterID != "" {
 		clusterInfo, err := clustersAPI.Get(clusterID)
+		if common.IsMissing(err) {
+			if instanceProfile == "" {
+				return fmt.Errorf("instance profile is required to re-create mounting cluster")
+			}
+			return mountS3ViaProfileAndSetClusterID(clustersAPI, instanceProfile, d)
+		}
 		if err != nil {
 			return err
 		}
@@ -69,11 +75,16 @@ func preprocessS3MountGeneric(ctx context.Context, s map[string]*schema.Schema, 
 			return fmt.Errorf("cluster %s must have EC2 instance profile attached", clusterID)
 		}
 	} else if instanceProfile != "" {
-		cluster, err := GetOrCreateMountingClusterWithInstanceProfile(clustersAPI, instanceProfile)
-		if err != nil {
-			return err
-		}
-		return d.Set("cluster_id", cluster.ClusterID)
+		return mountS3ViaProfileAndSetClusterID(clustersAPI, instanceProfile, d)
 	}
 	return nil
+}
+
+func mountS3ViaProfileAndSetClusterID(clustersAPI clusters.ClustersAPI, 
+	instanceProfile string, d *schema.ResourceData) error {
+	cluster, err := GetOrCreateMountingClusterWithInstanceProfile(clustersAPI, instanceProfile)
+	if err != nil {
+		return fmt.Errorf("mount via profile: %w", err)
+	}
+	return d.Set("cluster_id", cluster.ClusterID)
 }

--- a/storage/s3_test.go
+++ b/storage/s3_test.go
@@ -1,0 +1,106 @@
+package storage
+
+import (
+	"context"
+	"testing"
+
+	"github.com/databrickslabs/terraform-provider-databricks/clusters"
+	"github.com/databrickslabs/terraform-provider-databricks/common"
+	"github.com/databrickslabs/terraform-provider-databricks/qa"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestPreprocessS3MountOnDeletedClusterNoInstanceProfileSpecifiedError(t *testing.T) {
+	qa.HTTPFixturesApply(t, []qa.HTTPFixture{
+		{
+			Method:   "GET",
+			Resource: "/api/2.0/clusters/get?cluster_id=removed-cluster",
+			Status:   404,
+			Response: common.NotFound("cluster deleted"),
+		},
+	}, func(ctx context.Context, client *common.DatabricksClient) {
+		r := ResourceMount()
+		d := r.TestResourceData()
+		d.Set("uri", "s3://bucket")
+		d.Set("cluster_id", "removed-cluster")
+		err := preprocessS3MountGeneric(ctx, r.Schema, d, client)
+		assert.EqualError(t, err, "instance profile is required to re-create mounting cluster")
+	})
+}
+
+func TestPreprocessS3MountOnDeletedClusterWorks(t *testing.T) {
+	qa.HTTPFixturesApply(t, []qa.HTTPFixture{
+		{
+			Method:   "GET",
+			Resource: "/api/2.0/clusters/get?cluster_id=removed-cluster",
+			Status:   404,
+			Response: common.NotFound("cluster deleted"),
+		},
+		{
+			Method:   "GET",
+			Resource: "/api/2.0/clusters/list",
+			Response: clusters.ClusterList{
+				Clusters: []clusters.ClusterInfo{},
+			},
+		},
+		{
+			ReuseRequest: true,
+			Method:       "GET",
+			Resource:     "/api/2.0/clusters/spark-versions",
+		},
+		{
+			ReuseRequest: true,
+			Method:       "GET",
+			Resource:     "/api/2.0/clusters/list-node-types",
+		},
+		{
+			Method:   "POST",
+			Resource: "/api/2.0/clusters/create",
+			ExpectedRequest: clusters.Cluster{
+				CustomTags: map[string]string{
+					"ResourceClass": "SingleNode",
+				},
+				ClusterName:  "terraform-mount-s3-access",
+				SparkVersion: "7.3.x-scala2.12",
+				NumWorkers:   0,
+				NodeTypeID:   "i3.xlarge",
+				AwsAttributes: &clusters.AwsAttributes{
+					Availability:       "SPOT",
+					InstanceProfileArn: "arn:aws:iam::1234567:instance-profile/s3-access",
+				},
+				AutoterminationMinutes: 10,
+				SparkConf: map[string]string{
+					"spark.databricks.cluster.profile": "singleNode",
+					"spark.master":                     "local[*]",
+					"spark.scheduler.mode":             "FIFO",
+				},
+			},
+			Response: clusters.ClusterID{
+				ClusterID: "new-cluster",
+			},
+		},
+		{
+			Method:   "GET",
+			Resource: "/api/2.0/clusters/get?cluster_id=new-cluster",
+			Response: clusters.ClusterInfo{
+				ClusterID:    "new-cluster",
+				State:        "RUNNING",
+				StateMessage: "created",
+			},
+		},
+	}, func(ctx context.Context, client *common.DatabricksClient) {
+		r := ResourceMount()
+		d := r.TestResourceData()
+		d.MarkNewResource()
+		common.StructToData(GenericMount{
+			URI:       "s3://bucket",
+			ClusterID: "removed-cluster",
+			S3: &S3IamMount{
+				InstanceProfile: "arn:aws:iam::1234567:instance-profile/s3-access",
+			},
+		}, r.Schema, d)
+		err := preprocessS3MountGeneric(ctx, r.Schema, d, client)
+		assert.NoError(t, err)
+		assert.Equal(t, "new-cluster", d.Get("cluster_id"))
+	})
+}

